crl/validator.go

```diff
@@ -20,346 +20,388 @@ package crl
+	revoked map[string]bool
+	// issuer of the CRL found at this endpoint. Multiple issuers could indicate MITM attack, or re-use of an endpoint.
+	issuer *x509.Certificate
+	// lastUpdated is the timestamp that list was last updated. (When this instance was of revocationList was created)
+	lastUpdated time.Time
 }
 
-// NewValidatorWithHTTPClient returns a new instance with a pre-configured HTTP client
-func NewValidatorWithHTTPClient(certificates []*x509.Certificate, httpClient *http.Client) Validator {
-	certMap := map[string]*x509.Certificate{}
-
-	for _, certificate := range certificates {
-		certMap[certificate.Subject.String()] = certificate
-	}
-
-	return &validator{
-		httpClient:   httpClient,
-		bitSet:       NewBitSet(defaultBitSetSize),
-		certificates: certMap,
-		lists:        map[string]*x509.RevocationList{},
+func newRevocationList(cert *x509.Certificate) *revocationList {
+	return &revocationList{
+		list:    new(x509.RevocationList),
+		revoked: make(map[string]bool, 0),
+		issuer:  cert,
 	}
 }
 
-func (v *validator) setRevoked(issuer string, serialNumber *big.Int) {
-	bitNum := hash(issuer, serialNumber) % int64(v.bitSet.Len())
-
-	v.bitSet.Set(bitNum)
+// New returns a new CRL validator.
+func New(truststore []*x509.Certificate) Validator {
+	return newValidatorWithHTTPClient(truststore, http.DefaultClient)
 }
 
-// IsRevoked checks whether the certificate was revoked or not
-func (v *validator) IsRevoked(issuer string, serialNumber *big.Int) bool {
-	bitNum := hash(issuer, serialNumber) % int64(v.bitSet.Len())
-	if !v.bitSet.IsSet(bitNum) {
-		return false
+// NewValidatorWithHTTPClient returns a new instance with a pre-configured HTTP client
+func newValidatorWithHTTPClient(certificates []*x509.Certificate, client *http.Client) *validator {
+	val := &validator{
+		httpClient: client,
+		truststore: map[string]*x509.Certificate{},
 	}
-
-	v.listsLock.RLock()
-	defer v.listsLock.RUnlock()
-
-	for _, list := range v.lists {
-		listIssuerName := list.Issuer.String()
-
-		for _, cert := range list.RevokedCertificates {
-			if listIssuerName == issuer &&
-				cert.SerialNumber.Cmp(serialNumber) == 0 {
-				return true
-			}
+	// add truststore
+	for _, certificate := range certificates {
+		if _, ok := val.truststore[certificate.Subject.String()]; ok {
+			// skip duplicates
+			continue
 		}
+		val.truststore[certificate.Subject.String()] = certificate
 	}
-
-	return false
-}
-
-func (v *validator) updateCRL(endpoint string) error {
-	v.listsLock.RLock()
-	crl, ok := v.lists[endpoint]
-	v.listsLock.RUnlock()
-
-	if !ok || crlHasExpired(crl, nowFunc()) {
-		return v.downloadCRL(endpoint)
+	// add CRL distribution points
+	for _, certificate := range val.truststore {
+		issuer, ok := val.truststore[certificate.Issuer.String()]
+		if !ok {
+			err := fmt.Errorf("certificate's issuer is not in the trust store: subject=%s, issuer=%s", certificate.Subject.String(), certificate.Issuer.String())
+			panic(err)
```

Comment:
The problem is that (our) constructors almost never return an error. Giving this one an error limits the locations where this could be called from. The location the constructor is called from in the connection manager for example cannot handle errors (inside a `ConfigOption`). It panics in `grpc/config.go` when `SetValidatePeerCertificateFunc` fails for the same reason. We can add errors all the way upstream until the error can be handled gracefully, but that just makes the code very complicated. So this seemed the lesser of two evils to me, but I'm open to discus this.

Reply:
Maybe it should be passed to the gRPCConnectionManager?

Reply:
panics can not happen due to user error, only programming error. added comments

Reply:
@reinkrul see [don't panic](https://github.com/nuts-foundation/nuts-node/pull/2011/commits/49d2e185010de10b8d36ebcc6c169d80801e6ea5) commit for changes needed to do this. Makes the code better imo, maybe just needs to be in a separate issue.